docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Ranking/Ranking.ttinclude

```diff
@@ -0,0 +1,108 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.ML;
+using Microsoft.ML.Data;
+<# if (TrainerOptions != null) { #>
+<#=OptionsInclude#>
+<# } #>
+
+namespace Samples.Dynamic.Trainers.Ranking
+{
+    public static class <#=ClassName#>
+    {<#=Comments#>
+        public static void Example()
+        {
+            // Create a new context for ML.NET operations. It can be used for exception tracking and logging, 
+            // as a catalog of available operations and as the source of randomness.
+            // Setting the seed to a fixed number in this example to make outputs deterministic.
+            var mlContext = new MLContext(seed: 0);
+
+            // Create a list of training data points.
+            var dataPoints = GenerateRandomDataPoints(1000);
+
+            // Convert the list of data points to an IDataView object, which is consumable by ML.NET API.
+            var trainingData = mlContext.Data.LoadFromEnumerable(dataPoints);
+
+<# if (TrainerOptions == null) { #>
+            // Define the trainer.
+            var pipeline = mlContext.Ranking.Trainers.<#=Trainer#>();
+<# } else { #>
+            // Define trainer options.
+            var options = new <#=TrainerOptions#>;
+
+            // Define the trainer.
+            var pipeline = mlContext.Ranking.Trainers.<#=Trainer#>(options);
+<# } #>
+
+            // Train the model.
+            var model = pipeline.Fit(trainingData);
+
+            // Create testing data. Use different random seed to make it different from training data.
+            var testData = mlContext.Data.LoadFromEnumerable(GenerateRandomDataPoints(500, seed:123));
+
+            // Run the model on test data set.
+            var transformedTestData = model.Transform(testData);
+
+            // Convert IDataView object to a list.
+            var predictions = mlContext.Data.CreateEnumerable<Prediction>(transformedTestData, reuseRowObject: false).ToList();
+
+            // Look at 5 predictions
+            foreach (var p in predictions.Take(5))
+                Console.WriteLine($"Label: {p.Label}, Score: {p.Score}");
+
+            <#=ExpectedOutputPerInstance#>
+
+            // Evaluate the overall metrics
+            var metrics = mlContext.Ranking.Evaluate(transformedTestData);
+            PrintMetrics(metrics);
+            
+            <#=ExpectedOutput#>
+        }
+
+        private static IEnumerable<DataPoint> GenerateRandomDataPoints(int count, int seed = 0, int groupSize = 10)
+        {
+            var random = new Random(seed);
+            float randomFloat() => (float)random.NextDouble();
+            for (int i = 0; i < count; i++)
+            {
+                var label = random.Next(0, 5);
```

Comment:
>var label = random.Next(0, 5) [](start = 16, length = 29)

isn't label = 0 while it uint/key means this is a missing label?

Reply:
So the underlying value can be 0. When reading as a `KeyType` there is mapping that will shift everything by +1. So that 0 maps to 1.
I looked at the `IDataView` that was generated and the label values go from 1 to 5. 

---
In reply to: [275928728](https://github.com/dotnet/machinelearning/pull/3338#discussion_r275928728) [](ancestors = 275928728)